File: 2º Semestre/AEDs II/Aula Prática/Exercícios Práticos/Aula 2016_08_29/Insercao.java
/**
 * Algoritmo de ordenacao por insercao
 * @author Max do Val Machado
 * @version 2 01/2015
 */

class Insercao extends Geracao {

	/**
	 * Construtor.
	 */
	public Insercao(){
		super();
	}


	/**
	 * Construtor.
	 * @param int tamanho do array de numeros inteiros.
	 */
	public Insercao(int tamanho){
		super(tamanho);
	}


	/**
	 * Algoritmo de ordenacao por insercao.
	 */
	public static void insercao() {
		for (int i = 1; i < n; i++) {
			int tmp = array[i];
			int j = i - 1;

			while ((j >= 0) && (array[j] > tmp)) {
				array[j + 1] = array[j];
				j--;
			}
			array[j + 1] = tmp;
		}
	}

	/**
	 * Algoritmo de ordenacao parcial por insercao.
	 * @params k int
	 */
	public static void insercao (int k){
		int j;
		for (int i = 1; i < n; i++) {
			int tmp = array[i];
			if(i > k){
				j = k;
				if(tmp < array[k]){
					array[i] = array[k];
				}
			} else{
				j = i-1;
			}
			while (j >= 0 && tmp < array[j]) {
				if(j < k){
					array[j + 1] = array[j];
				}
				j--;
			}
			if(j < k){ 
				array[j + 1] = tmp;
			}
		}
	}

	public static void main(String[] args) {
		Insercao insercao = new Insercao(20);
		insercao.aleatorio();
		insercao.mostrar();

		long comeco = now();
		insercao.insercao(5);
		long fim = now();

		insercao.mostrar();
		System.out.println("Tempo para ordenar parcialmente: " + (fim-comeco)/1000.0 + " s.");


	}
}

File: 2º Semestre/AEDs II/Aula Prática/Fonte/U4_ordenacao/insercao.c
/**
 * Metodo de ordenacao por insercao
 * @author Max do Val Machado
 * @version 2 01/2015
 */
#include "geracao.h"

// Algoritmo de ordenacao
void insertionSort() {
   for (i = 1; i < n; i++) {
      int tmp = array[i];
         int j = i - 1;
         
         while ((j >= 0) && (array[j] > tmp)) {
            array[j + 1] = array[j];
               j--;
         }
      array[j + 1] = tmp;
         
   }
}

int main() {
   srand(time(NULL));
   decrescente();
   //mostrar();		
   clock_t comeco = clock();
   insertionSort();
   clock_t fim = clock();
   int total = (clock() - comeco) / CLOCKS_PER_SEC / 1000;

   mostrar();
   printf("Tempo para ordenar: %d ms.", total);
}
